### PR TITLE
Check if domain name is NULL when constructing util info.

### PR DESCRIPTION
**Pull request by bturrubiates**
If the domain name is NULL, this causes a seg fault in strdup.

Hit this issue because the getinfo test in fabtests makes a request using FI_VERSION(1, 0) and for compatibility reasons, usNIC returns a NULL domain name. 

Signed-off-by: Ben Turrubiates <bturrubi@cisco.com>

### CODE REVIEW

**Comment by shefty** [comment]
Returning a null domain name sounds more like a bug to me.  Surely an app is dependent on getting buggy results from fi_getinfo?

**Comment by bturrubiates** [comment]
> Returning a null domain name sounds more like a bug to me. Surely an app is dependent on getting buggy results from fi_getinfo?

Yes, but there are releases out there with this behavior and there was some discussion about this a while ago and the thing that ended up getting implemented was fixing it in future releases but returning the old behavior when an app uses an old FI_VERSION value.

**Comment by shefty** [comment]
That's arguing to keep buggy code because it's already been released.  Heck, you might as well discard all bug fixes since 1.0 then.  :P  The behavior you're describing is unique to usnic.  Other providers did the correct thing.  And now this change is allowing other providers to make the same mistake in the future.  I don't like the reason for this change at all.  It comes across as taking the wrong approach.

**Comment by bturrubiates** [comment]
> That's arguing to keep buggy code because it's already been released. Heck, you might as well discard all bug fixes since 1.0 then. :P The behavior you're describing is unique to usnic. Other providers did the correct thing. And now this change is allowing other providers to make the same mistake in the future. I don't like the reason for this change at all. It comes across as taking the wrong approach.

Well the decision at the time was that requesting an older version with a newer library should reflect the same behavior as an older version. Before version 1.4, there wasn't a common meaning for those fields between providers (this is when usNIC and sockets decided to use CIDR notation). 

@jsquyres Adding you to the conversation, thoughts?

**Comment by jsquyres** [comment]
I have to agree with @bturrubiates here -- you can't say that this is buggy behavior because there was no standard for the names before v1.4.  So by definition, what usnic is doing is *not* buggy when an API version of < 1.4 is given.

Can the `version` parameter be used here to restrict Ben's new (old) behavior to only when it is < v1.4?  That would seem to honor the letter of the law (i.e., behave just like the older version) but also prevent newer versions from newly-defined-Bad-Behavior (i.e., don't let them return NULL).

**Comment by shefty** [comment]
There was no agreement on the format of the name, but arguing that NULL is a correct identifier seems like a stretch.  Regardless, if we assume the intent of having no format, then I don't see why it can't change.  What application is specifically looking for NULL fields here, let along one looking based on the version?

**Comment by jsquyres** [comment]
The usNIC component in Open MPI requests libfabric API version 1.3 or 1.4: https://github.com/open-mpi/ompi/blob/master/opal/mca/btl/usnic/btl_usnic_component.c#L671-L681

It's not that Open MPI looks for the domain name to be NULL, it just chooses where to get the Linux device name of the underlying usNIC device (i.e., the fabric or the device name): https://github.com/open-mpi/ompi/blob/master/opal/mca/btl/usnic/btl_usnic_component.c#L823-L828

Our point is not that Open MPI will break if we change this behavior (i.e., that the domain name in API <=v1.3 suddenly started returning non-NULL), but rather that changing the behavior of an old API version -- that was previously unspecified behavior -- is potentially dangerous.  

Granted: it's probably not a tragedy if usnic API <=v1.3 starts returning a `""` instead of NULL.  But the principle of the issue is more important here: unless there's a very, very good reason, we should leave the old behavior the same so as to avoid unintended consequences.

**Comment by shefty** [comment]
If it breaks an app, I agree we need to maintain the same behavior.  @bturrubiates - can you update the patch as Jeff suggested to include a check for the version?

**Comment by bturrubiates** [comment]
> If it breaks an app, I agree we need to maintain the same behavior. @bturrubiates - can you update the patch as Jeff suggested to include a check for the version?

Sure, so to be clear you want the NULL check only on versions > 1.3?

**Comment by shefty** [comment]
If the version is <= 1.3 (or whatever version you need), allow NULL.  For version > 1.3, assert() that the name is not NULL.

**Comment by bturrubiates** [comment]
> If the version is <= 1.3 (or whatever version you need), allow NULL. For version > 1.3, assert() that the name is not NULL.

Updated with an assert.

File: prov/util/src/util_attr.c [hunk 1]
```diff
@@ -230,11 +231,23 @@ static int ofi_info_to_util(uint32_t version, const struct fi_provider *prov,
...
-		FI_WARN(prov, FI_LOG_FABRIC,
-			"Unable to allocate domain name\n");
-		goto err;
+	/* Release 1.4 brought standardized domain names across IP based
+	 * providers. Before this release, the usNIC provider would return a
+	 * NULL domain name from fi_getinfo. For compatibility reasons, allow a
+	 * NULL domain name when apps are requesting version < 1.4.
+	 */
+	if (FI_VERSION_GE(version, FI_VERSION(1, 4)))
+		assert(core_info->domain_attr->name);
```
Comment (shefty) [inline prov/util/src/util_attr.c]: You can move the if check into the assert.  E.g. assert(FI_VERSION_LT(1.4) || name)

**Comment by bturrubiates** [comment]
@shefty Updated.

**Pull request merged**